pkg/gadgets/profile/tcprtt/tracer/tracer.go

```diff
@@ -0,0 +1,129 @@
+
+package tracer
+
+import (
+	"encoding/json"
+	"fmt"
+	"unsafe"
+
+	"github.com/cilium/ebpf"
+	"github.com/cilium/ebpf/link"
+
+	gadgetcontext "github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-context"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets/profile/tcprtt/types"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/logger"
+)
+
+//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -target $TARGET -type hist -cc clang tcpRTT ./bpf/tcprtt.bpf.c -- -I./bpf/ -I../../../../${TARGET}
+
+type Tracer struct {
+	objs                tcpRTTObjects
+	tcpRcvEstKprobeLink link.Link
+	tcpRcvEstFentryLink link.Link
+
+	logger logger.Logger
+}
+
+func (t *Tracer) RunWithResult(gadgetCtx gadgets.GadgetContext) ([]byte, error) {
+	t.logger = gadgetCtx.Logger()
+
+	defer t.close()
+	if err := t.install(); err != nil {
+		return nil, fmt.Errorf("installing tracer: %w", err)
+	}
+
+	gadgetcontext.WaitForTimeoutOrDone(gadgetCtx)
+
+	result, err := t.collectResult()
+	if err != nil {
+		return nil, fmt.Errorf("collecting result: %w", err)
+	}
+	return result, nil
+}
+
+func (t *Tracer) collectResult() ([]byte, error) {
+	if t.objs.Hists == nil {
+		t.logger.Debugf("No result to collect")
+		return nil, nil
+	}
+	histMap := t.objs.Hists
+
+	// TODO: Iterate over keys when we support showing histogram by address
+	var key uint32
+	if err := histMap.NextKey(nil, unsafe.Pointer(&key)); err != nil {
+		return []byte{}, fmt.Errorf("getting next key: %w", err)
+	}
+
+	t.logger.Debugf("Collecting result for key %d", key)
+
+	hist := tcpRTTHist{}
+	if err := histMap.Lookup(key, unsafe.Pointer(&hist)); err != nil {
+		return []byte{}, fmt.Errorf("getting histogram for key %d: %w", key, err)
+	}
+
+	// TODO: Support milliseconds
+	report := types.NewReport("usecs", hist.Slots[:])
+
+	// TODO: Support extension summary (average)
+	if hist.Cnt > 0 {
+		t.logger.Debugf("[AVG %llu]", hist.Latency/hist.Cnt)
+	} else {
+		t.logger.Debugf("[AVG 0]")
+	}
+
+	return json.Marshal(report)
+}
+
+func (t *Tracer) close() {
+	t.tcpRcvEstKprobeLink = gadgets.CloseLink(t.tcpRcvEstKprobeLink)
+	t.tcpRcvEstFentryLink = gadgets.CloseLink(t.tcpRcvEstFentryLink)
+
+	t.objs.Close()
+}
+
+func (t *Tracer) install() error {
+	var spec *ebpf.CollectionSpec
+
+	spec, err := loadTcpRTT()
+	if err != nil {
+		return fmt.Errorf("loading specs: %w", err)
+	}
+
+	if err := spec.LoadAndAssign(&t.objs, nil); err != nil {
+		return fmt.Errorf("loading ebpf program: %w", err)
+	}
+
+	tcpRcvEstKprobeLink, err := link.Kprobe("tcp_rcv_established", t.objs.IgTcprcvestKp, nil)
+	if err != nil {
+		return fmt.Errorf("attaching kprobe: %w", err)
+	}
+	t.tcpRcvEstKprobeLink = tcpRcvEstKprobeLink
+
+	tcpRcvEstFentryLink, err := link.AttachTracing(link.TracingOptions{Program: t.objs.IgTcprcvestFe})
+	if err != nil {
+		return fmt.Errorf("attaching fentry: %w", err)
+	}
+	t.tcpRcvEstFentryLink = tcpRcvEstFentryLink
```

Comment:
I think only one of those programs has to be attached. Fentry for newer kernels and Kprobe for older ones. Btw, in some gadges we only support kprobes and fsslower is future work:

https://github.com/inspektor-gadget/inspektor-gadget/blob/1b0eb1d2789f462e9acb213f3d9d1595cb46eb98/pkg/gadgets/trace/fsslower/tracer/bpf/fsslower.bpf.c#L184

The issue is that (last time I checked) there is not a way in cilium/ebpf to disable loading a program, so in all cases we'll have both programs around (even if we don't attach them). This is solved in BCC by using `bpf_program__set_autoload` to disable programs that are not needed. 

I'll propose to only support kprobes for now until we investigate (and hopefully) implement that feature in cilium/ebpf.

Reply:
ref https://github.com/inspektor-gadget/inspektor-gadget/issues/1566

Reply:
I commented the fentry one.